File: internal/actions/webhook.go
package actions

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"github.com/google/uuid"

	"github.com/openforms/openforms/internal/jobs"
	"github.com/openforms/openforms/internal/submissions"
)

// Sign returns "sha256=<hex HMAC-SHA256(secret, body)>".
func Sign(secret string, body []byte) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

// webhookSubmission mirrors the Submission JSON of spec §7.3.
type webhookSubmission struct {
	ID          uuid.UUID        `json:"id"`
	Form        string           `json:"form"`
	FormVersion int              `json:"formVersion"`
	State       string           `json:"state"`
	StateLabel  string           `json:"stateLabel"`
	Terminal    bool             `json:"terminal"`
	Data        map[string]any   `json:"data"`
	Fields      map[string]any   `json:"fields"`
	Assignee    *webhookAssignee `json:"assignee"`
	CreatedAt   time.Time        `json:"createdAt"`
	UpdatedAt   time.Time        `json:"updatedAt"`
}

type webhookAssignee struct {
	ID    uuid.UUID `json:"id"`
	Name  string    `json:"name"`
	Email string    `json:"email"`
}

type webhookTransition struct {
	Key   string `json:"key"`
	Label string `json:"label"`
	From  string `json:"from"`
	To    string `json:"to"`
}

type webhookForm struct {
	Slug  string `json:"slug"`
	Title string `json:"title"`
}

type webhookBody struct {
	Event      string             `json:"event"`
	Submission webhookSubmission  `json:"submission"`
	Transition *webhookTransition `json:"transition"`
	Form       webhookForm        `json:"form"`
}

func toWebhookSubmission(s submissions.Submission) webhookSubmission {
	var assignee *webhookAssignee
	if s.AssigneeID != nil {
		assignee = &webhookAssignee{ID: *s.AssigneeID, Name: s.AssigneeName, Email: s.AssigneeEmail}
	}
	return webhookSubmission{
		ID: s.ID, Form: s.FormSlug, FormVersion: s.FormVersion,
		State: s.State, StateLabel: s.StateLabel, Terminal: s.Terminal,
		Data: nonNil(s.Data), Fields: nonNil(s.Fields), Assignee: assignee,
		CreatedAt: s.CreatedAt.UTC(), UpdatedAt: s.UpdatedAt.UTC(),
	}
}

func (r *runner) webhook(ctx context.Context, job jobs.Job) error {
	ac, err := r.load(ctx, job)
	if err != nil {
		return err
	}
	target := ac.payload.Action.URL
	u, err := url.Parse(target)
	if err != nil || (u.Scheme != "http" && u.Scheme != "https") || u.Host == "" {
		return jobs.Permanent(fmt.Errorf("webhook: invalid url %q", target))
	}

	body := webhookBody{
		Event:      "submission.transitioned",
		Submission: toWebhookSubmission(ac.sub),
		Form:       webhookForm{Slug: ac.form.Slug, Title: ac.form.Title},
	}
	if ac.payload.Trigger == TriggerSubmit {
		body.Event = "submission.created"
	} else if ac.transition != nil {
		body.Transition = &webhookTransition{Key: ac.transition.Key, Label: ac.transition.Label, From: ac.fromState, To: ac.transition.To}
	}
	raw, err := json.Marshal(body)
	if err != nil {
		return jobs.Permanent(err)
	}

	reqCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, target, bytes.NewReader(raw))
	if err != nil {
		return jobs.Permanent(err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "openforms-webhook/1")
	req.Header.Set("X-OpenForms-Event", body.Event)
	req.Header.Set("X-OpenForms-Delivery", strconv.FormatInt(job.ID, 10))
	if secret := r.d.Config.WebhookSecret; secret != "" {
		req.Header.Set("X-OpenForms-Signature", Sign(secret, raw))
	}

	resp, err := r.d.HTTPClient.Do(req)
	if err != nil {
		return fmt.Errorf("webhook %s: %w", target, err)
	}
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 64<<10))
	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		e := fmt.Errorf("webhook %s responded %d", target, resp.StatusCode)
		if resp.StatusCode >= 400 && resp.StatusCode < 500 &&
			resp.StatusCode != http.StatusRequestTimeout && resp.StatusCode != http.StatusTooManyRequests {
			return jobs.Permanent(e)
		}
		return e
	}
	return r.recordSuccess(ctx, ac)
}

File: src/openforms/server/actions/webhook.py
"""Webhook delivery: signed JSON POST, 10 s timeout; 4xx (except 408/429) is permanent."""

from __future__ import annotations

import hashlib
import hmac
import json
from typing import TYPE_CHECKING, Any

import httpx

from ...definition.common import is_http_url
from ..api.serialize import ts
from ..models.submissions import Submission
from ..services.jobs import Job, Permanent
from .payload import TRIGGER_SUBMIT

if TYPE_CHECKING:
    from .runner import Runner


def sign(secret: str, body: bytes) -> str:
    """``sha256=<hex HMAC-SHA256(secret, body)>``."""
    return "sha256=" + hmac.new(secret.encode(), body, hashlib.sha256).hexdigest()


def webhook_submission(s: Submission) -> dict[str, Any]:
    return {
        "id": str(s.id),
        "form": s.form_slug,
        "formVersion": s.form_version,
        "state": s.state,
        "stateLabel": s.state_label,
        "terminal": s.terminal,
        "data": s.data or {},
        "fields": s.fields or {},
        "assignee": None
        if s.assignee_id is None
        else {"id": str(s.assignee_id), "name": s.assignee_name, "email": s.assignee_email},
        "createdAt": ts(s.created_at),
        "updatedAt": ts(s.updated_at),
    }


async def run_webhook(r: Runner, job: Job) -> None:
    ac = await r.load(job)
    target = ac.payload.action.url
    if not is_http_url(target):
        raise Permanent(ValueError(f"webhook: invalid url {target!r}"))
    body: dict[str, Any] = {
        "event": "submission.transitioned",
        "submission": webhook_submission(ac.sub),
        "transition": None,
        "form": {"slug": ac.form.slug, "title": ac.form.title},
    }
    if ac.payload.trigger == TRIGGER_SUBMIT:
        body["event"] = "submission.created"
    elif ac.transition is not None:
        body["transition"] = {
            "key": ac.transition.key,
            "label": ac.transition.label,
            "from": ac.from_state,
            "to": ac.transition.to,
        }
    raw = json.dumps(body, separators=(",", ":"), ensure_ascii=False).encode()
    headers = {
        "Content-Type": "application/json",
        "User-Agent": "openforms-webhook/1",
        "X-OpenForms-Event": body["event"],
        "X-OpenForms-Delivery": str(job.id),
    }
    if r.d.settings.webhook_secret:
        headers["X-OpenForms-Signature"] = sign(r.d.settings.webhook_secret, raw)
    try:
        resp = await r.http.post(target, content=raw, headers=headers, timeout=10.0)
    except httpx.HTTPError as e:
        raise RuntimeError(f"webhook {target}: {e!r}") from e
    if not 200 <= resp.status_code <= 299:
        err = RuntimeError(f"webhook {target} responded {resp.status_code}")
        if 400 <= resp.status_code < 500 and resp.status_code not in (408, 429):
            raise Permanent(err)
        raise err
    await r.record_success(ac)
